volatility/framework/plugins/windows/hashdump.py

```diff
@@ -0,0 +1,346 @@
+    208, 208, 211, 211, 213, 213, 214, 214, 217, 217, 218, 218, 220, 220, 223, 223,
+    224, 224, 227, 227, 229, 229, 230, 230, 233, 233, 234, 234, 236, 236, 239, 239,
+    241, 241, 242, 242, 244, 244, 247, 247, 248, 248, 251, 251, 253, 253, 254, 254
+    ]
+
+    # Permutation matrix for boot key
+    p = [ 0x8, 0x5, 0x4, 0x2, 0xb, 0x9, 0xd, 0x3,
+        0x0, 0x6, 0x1, 0xc, 0xe, 0xa, 0xf, 0x7 ]
+
+    # Constants for SAM decrypt algorithm
+    aqwerty = b"!@#$%^&*()qwertyUIOPAzxcvbnmQQQQQQQQQQQQ)(*@&%\0"
+    anum = b"0123456789012345678901234567890123456789\0"
+    antpassword = b"NTPASSWORD\0"
+    almpassword = b"LMPASSWORD\0"
+    lmkey = b"KGS!@#$%"
+
+    empty_lm = "aad3b435b51404eeaad3b435b51404ee"
+    empty_nt = "31d6cfe0d16ae931b73c59d7e0c089c0"
+
+    #empty treegrid object to return in case of error
+    none_obj = (0,('','','',''))
+
+    #helper method to find a particular key in a hive
+    def open_key(self, root, key):
+        if key == []:
+            return root
+
+        keyname = key.pop(0)
+        for s in root.get_subkeys():        
+            if s.get_name().upper() == keyname.upper():
+                    return self.open_key(s, key)
+        return None
+
+    def get_user_keys(self, samhive):
+        user_key_path = ["SAM", "Domains", "Account", "Users"]
+
+        root = samhive.root_cell_offset
+        if not root:
+            return []
+
+        user_key = self.open_key(samhive.get_node(root), user_key_path)
+        if not user_key:
+            return []
+        return [k for k in user_key.get_subkeys() if k.Name != "Names"]
+
+    def get_bootkey(self, syshive):
+        cs =1
+        lsa_base = ["ControlSet{0:03}".format(cs), "Control", "Lsa"]
+        lsa_keys = ["JD", "Skew1", "GBG", "Data"]
+        root = syshive.root_cell_offset 
+        if not root:
+            return None
+
+        lsa = self.open_key(syshive.get_node(root), lsa_base)
+
+
+        if not lsa:
+            return None
+
+        bootkey = ""
+
+        for lk in lsa_keys:
+            key = self.open_key(lsa, [lk])
+
+            class_data = syshive.read(key.Class+4, key.ClassLength)
+
+            if class_data == None:
+                return ""
+            bootkey += class_data.decode('utf-16-le')
+            
+
+
+        bootkey_str=''.join([chr(int(''.join(c), 16)) for c in zip(bootkey[0::2],bootkey[1::2])])
+        bootkey_scrambled = ""
+
+        for i in range(len(bootkey_str)):
+            bootkey_scrambled += bootkey_str[self.p[i]]
+        return bootkey_scrambled
+
+    def get_hbootkey(self, samhive, bootkey):
+        sam_account_path = ["SAM", "Domains", "Account"]
+
+        if not bootkey:
+            return None
+
+        root = samhive.root_cell_offset
+        if not root:
+            return None
+
+        sam_account_key = self.open_key(samhive.get_node(root), sam_account_path)
+        if not sam_account_key:
+            return None
+        
+
+        F = None
```

Comment:
We try to use descriptive variable names, otherwise such things get extremely difficult to follow down the line, even if this is literally called `F_value` something a bit more than just `F` that helps people know what it is would be useful...